apps/tangle-dapp/data/liquidStaking/helper.ts

```diff
@@ -0,0 +1,322 @@
+import { Option } from '@polkadot/types';
+import { AccountId32 } from '@polkadot/types/interfaces';
+import { BN, BN_ZERO } from '@polkadot/util';
+
+import {
+  PalletDappStakingV3ContractStakeAmount,
+  PalletDappStakingV3DAppInfo,
+} from '../../types/astarDappStaking';
+import { getApiPromise } from '../../utils/polkadot';
+
+export const fetchValidators = async (
+  rpcEndpoint: string,
+): Promise<AccountId32[]> => {
+  const api = await getApiPromise(rpcEndpoint);
+  const validators = (await api.query.session.validators()).map((val) => val);
+  return validators;
+};
+
+export const fetchMappedIdentityNames = async (
+  rpcEndpoint: string,
+): Promise<Map<string, string>> => {
+  const api = await getApiPromise(rpcEndpoint);
+  const map = new Map<string, string>();
+
+  if (!api.query.identity) {
+    // throw new Error('identity pallet is not available in the runtime');
+    return map;
+  }
+
+  const identityNames = await api.query.identity.identityOf.entries();
+
+  identityNames.forEach((identity) => {
+    if (!identity[1].unwrap()[0]) {
+      return;
+    }
+
+    const validator = identity[0].args[0];
+    const info = identity[1].unwrap()[0].info;
+    let displayName = validator.toString();
+    if (info) {
+      const displayData = info['display'];
+      if (displayData.isNone) return null;
+      const displayDataObject: { raw?: string } = JSON.parse(
+        displayData.toString(),
+      );
+      if (displayDataObject.raw !== undefined) {
+        const hexString = displayDataObject.raw;
+        displayName = Buffer.from(hexString.slice(2), 'hex').toString('utf8');
+      }
+    }
+    map.set(validator.toString(), displayName);
+  });
+  return map;
+};
+
+export const fetchMappedValidatorsTotalValueStaked = async (
+  rpcEndpoint: string,
+): Promise<Map<string, BN>> => {
+  const api = await getApiPromise(rpcEndpoint);
+
+  if (!api.query.staking) {
+    throw new Error('staking pallet is not available in the runtime');
+  }
+
+  const activeEra = await api.query.staking.activeEra();
+  const activeEraIndex = activeEra.unwrap().index;
+  const validators = await fetchValidators(rpcEndpoint);
+  const stakerPromises = validators.map(async (val) => {
+    const erasStakers = await api.query.staking.erasStakersOverview(
+      activeEraIndex,
+      val,
+    );
+    const exposure = erasStakers.unwrap();
+    const totalStaked = exposure.total.unwrap().toBn() || BN_ZERO;
+    return {
+      validator: val.toString(),
+      totalStaked,
+    };
+  });
+  const stakers = await Promise.all(stakerPromises);
+  const map = new Map<string, BN>();
+  stakers.forEach((staker) => {
+    map.set(staker.validator, staker.totalStaked);
+  });
+  return map;
+};
```

Comment:
If possible, we should move to using subscriptions instead of requests because this approach may end up similar to the previous performance problems we've faced, for example, here there is an operation performed per validator `validators.map(async (val) => { ... await api.query.staking.erasStakersOverview() ... }`, so if there's 600+ validators, that could already be a pain point

Reply:
Removed all the async calls made inside the loops. Example:  Instead of calling `await api.query.staking.erasStakersOverview()` for each validator, now calling `await api.query.staking.erasStakersOverview.entries()` once which fetches all the era stakers overview at once and then use `Map`. Same for `fetchMappedValidatorsCommission` fucntion as well. After this no function does async call inside of a loop.